### PR TITLE
8310886: C2 SuperWord: Two nodes should be isomorphic if they are loop invariant but pinned at different nodes outside the loop

### PR DESCRIPTION
**Pull request by chhagedorn**
When currently hoisting a range check out of a loop in Loop Predication, we create two Hoisted Check Predicate (one for the init value and one for the last iteration value) together with two Template Assertion Predicate (one for the init value and one for the last iteration value). Any data dependency on a hoisted range check (for which `depends_only_on_test()` is true) is rewired to the second Template Assertion Predicate:

image
- `487 RangeCheck` + `496 RangeCheck`: Hoisted Check Predicate
- `405 If` + `416 Range Check`: Template Assertion Predicate
- `218 LoadI` rewired data dependency from hoisted range check inside the loop

When creating pre/main/post loops, we rewire all data dependencies found at Template Assertion Predicates which are part of the original loop body (and therefore cloned and part of the pre, main, and post loop) to the main and post loop. Currently, they all end up at the zero trip guard of the main loop and post loop (image shows the main loop zero trip guard):

image

When further unrolling the main loop, all cloned data dependencies stay at the zero trip guard. For example, after unrolling three times, we have all the cloned loads still at the zero trip guard:

image

The Java code for the graphs above is:
```
    static void test(int[] a, int[] b) {
        for (int i = 0; i < 1024; i+=2) {
            a[i+0] += b[i+0];
            a[i+1] += b[i+1];
        }
    }
```
We can apply Superword and vectorize the loads.

However, if we ever turn a main loop (before Superword) into a normal loop again and split it further (peel, unswitch, pre/main/post etc.). We would need to create new Assertion Predicates for the split loops and rewire the data dependencies to them. Both is currently not done and known to lead to failures. This should be fixed with JDK-8288981.

In JDK-8288981, I introduce a new `TemplateAssertionPredicateNodes` which replaces the two Template Assertion Predicates with their `Opaque4Nodes`. This means that for each hoisted range check, we create a single `TemplateAssertionPredicateNode`:

image

For the Java code in this example, we hoist four range checks and therefore create four `TemplateAssertionPredicateNodes`.

To fix the data dependency problem when splitting a loop, I keep the data dependencies at the cloned `TemplateAssertionPredicateNodes`. For example, the graph of this example with four hoisted checks looks like this after unrolling three times:

image

We've kept the data dependencies of all four hoisted range checks at their `TemplateAssertionPredicateNodes` instead of collecting them at the zero trip guard. 

But this now prevents Superword from vectorizing these nodes because the `isomorphic()` check of `a[i+0]` and `a[i+1]` fails: They do not have the same control node - even though both are pinned outside the loop. There is some special logic inside `isomorphic()`, though, which still allows the check to pass if both are pinned at range checks (not the case here because they are pinned at  `TemplateAssertionPredicateNodes`):
https://github.com/openjdk/jdk/blob/242eeaea47a259cab4ad2d4f0e055959e9870b8d/src/hotspot/share/opto/superword.cpp#L1280-L1292

or if a `MulAddS2I` node is involved (also not the case):
https://github.com/openjdk/jdk/blob/242eeaea47a259cab4ad2d4f0e055959e9870b8d/src/hotspot/share/opto/superword.cpp#L1293-L1305

Since both fail, we conclude that the loads are not isomorphic and we fail to vectorize the loads.

This additional special logic seems odd. We should always be able to vectorize pinned nodes as long as they are loop invariant (i.e. pinned outside the loop). I therefore propose to change this logic to only check for loop invariance, regardless of any pins. Additionally, Superword should be strong enough to prevent vectorization if there is problem with data dependencies inside and among the packs.

I've added an IR test to make sure that `MulAddS2I` nodes are still vectorized.

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310886](https://bugs.openjdk.org/browse/JDK-8310886): C2 SuperWord: Two nodes should be isomorphic if they are loop invariant but pinned at different nodes outside the loop (**Enhancement** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15842/head:pull/15842` \
`$ git checkout pull/15842`

Update a local copy of the PR: \
`$ git checkout pull/15842` \
`$ git pull https://git.openjdk.org/jdk.git pull/15842/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15842`

View PR using the GUI difftool: \
`$ git pr show -t 15842`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15842.diff">https://git.openjdk.org/jdk/pull/15842.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15842#issuecomment-1727748929)

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
Are you sure that VNNI instruction is used in your test? I don't see VNNI CPU feature check in the test.
Here is instruction:
https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/x86.ad#L8744

**Comment by chhagedorn** [comment]
> Are you sure that VNNI instruction is used in your test? I don't see VNNI CPU feature check in the test. Here is instruction: https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/x86.ad#L8744

You're right - I'm currently only checking if Superword succeeds and creates `MulAddVS2VINodes` but not if the matcher also matches `vmuladdaddS2I_reg` when VNNI is supported. I've pushed an update to add an additional IR rule which checks that `vmuladdaddS2I_reg` is matched and that the mach graph contains this node if VNNI is available.

**Review by vnkozlov — APPROVED** [review]
This looks good.

**Comment by chhagedorn** [comment]
Thanks Vladimir for your review!

File: test/hotspot/jtreg/compiler/lib/ir_framework/IRNode.java [hunk 1]
```diff
@@ -907,6 +907,22 @@ public class IRNode {
...
     }
 
+    public static final String MUL_ADD_S2I = PREFIX + "MUL_ADD_S2I" + POSTFIX;
+    static {
+        beforeMatchingNameRegex(MUL_ADD_S2I, "MulAddS2I");
+    }
+
+    public static final String MUL_ADD_VS2VI = PREFIX + "MUL_ADD_VS2VI" + POSTFIX;
+    static {
+        superWordNodes(MUL_ADD_VS2VI, "MulAddVS2VI");
```
Comment (eme64) [inline test/hotspot/jtreg/compiler/lib/ir_framework/IRNode.java:917]: Is there a reason this is not a `vectorNode`?

File: test/hotspot/jtreg/compiler/loopopts/superword/TestMulAddS2I.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    static final int ITER  = RANGE/2 - 1;
+
+    static short[] sArr1 = new short[RANGE];
+    static short[] sArr2 = new short[RANGE];
+    static final int[] GOLDEN;
+
+    static {
+        for (int i = 0; i < RANGE; i++) {
+            sArr1[i] = (short)(i + (1000 * AbstractInfo.getRandom().nextDouble()));
+            sArr2[i] = (short)(RANGE - i + (2000 * AbstractInfo.getRandom().nextDouble()));
```
Comment (eme64) [inline test/hotspot/jtreg/compiler/loopopts/superword/TestMulAddS2I.java:50]: Did you want to change this to `(short)AbstractInfo.getRandom().nextInt();`

File: test/hotspot/jtreg/compiler/loopopts/superword/TestMulAddS2I.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        }
+    }
+
+    @Test
+    @IR(applyIfCPUFeature = {"sse2", "true"}, applyIf = {"UseUnalignedLoadStores", "true"},
+        counts = {IRNode.MUL_ADD_S2I, "> 0", IRNode.MUL_ADD_VS2VI, "> 0"})
+    @IR(applyIfCPUFeature = {"sse2", "true"}, applyIf = {"UseUnalignedLoadStores", "false"},
+        failOn = {IRNode.MUL_ADD_VS2VI}, // Can only pack LoadS if UseUnalignedLoadStores is true (default if sse4.2)
+        counts = {IRNode.MUL_ADD_S2I, "> 0"})
+    @IR(applyIfCPUFeature = {"asimd", "true"}, applyIf = {"MaxVectorSize", "16"},
```
Comment (eme64) [inline test/hotspot/jtreg/compiler/loopopts/superword/TestMulAddS2I.java:83]: Why `"MaxVectorSize", "16"`?

**Review by eme64 — APPROVED** [review]
Looks good, I just had a few minor comments.

File: test/hotspot/jtreg/compiler/loopopts/superword/TestMulAddS2I.java (hunk 2 above)
Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/loopopts/superword/TestMulAddS2I.java:50]: Good idea, I will update this.

File: test/hotspot/jtreg/compiler/lib/ir_framework/IRNode.java (hunk 1 above)
Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/lib/ir_framework/IRNode.java:917]: First I've noticed that `MulAddVS2VI` is only created in Superword as far as I see. But now that [JDK-8310308](https://bugs.openjdk.org/browse/JDK-8310308) is in, I think I can switch to `vectorNode()`. Additionally, I think I need to revisit the rule for aarch64 to see if I can do better with the features from JDK-8310308. In the .ad file it says that it's only possible to use these nodes if the length is 16:
https://github.com/openjdk/jdk/blob/5350fd617390aaaedf8dd8821418c796cb1c38b3/src/hotspot/cpu/aarch64/aarch64_vector.ad#L182-L185

File: test/hotspot/jtreg/compiler/loopopts/superword/TestMulAddS2I.java (hunk 3 above)
Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/loopopts/superword/TestMulAddS2I.java]: The AD file enforces it:
https://github.com/openjdk/jdk/blob/5350fd617390aaaedf8dd8821418c796cb1c38b3/src/hotspot/cpu/aarch64/aarch64_vector.ad#L182-L185

I'm not sure if that limitation is too restrictive. I could be investigated in a separate RFE. But I think to ensure the correctness of this rule, we need this constraint for aarch64.

**Comment by chhagedorn** [comment]
I've pushed an update to use the features of JDK-8310308 and address the review comments.

**Review by eme64 — APPROVED** [review]
Looks even better now :)

**Review by vnkozlov — APPROVED** [review]
Last update is good.

**Comment by chhagedorn** [comment]
Thanks Vladimir for the re-review!

Testing looked good.

/integrate

**Pull request closed without merging**